### PR TITLE
[Data Explorer][Discover 2.0] Implement data fetch, index pattern, saved search for table vis

### PR DESCRIPTION
**Pull request by ananzh**
### Description
* Fetch data and feed to table vis by deangularizing controller methods
* Transfer legacy table to data grid Table
* Remove doc view links for now

### Issues Resolved
https://github.com/opensearch-project/OpenSearch-Dashboards/issues/4224

## Screenshot
Screenshot 2023-07-17 at 07 40 38

Screenshot 2023-07-17 at 07 46 52

### Check List

- [ ] All tests pass
  - [ ] `yarn test:jest`
  - [ ] `yarn test:jest_integration`
  - [ ] `yarn test:ftr`
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
- [ ] Update [CHANGELOG.md](./../CHANGELOG.md)
- [ ] Commits are signed per the DCO using --signoff

### CODE REVIEW

File: src/plugins/discover/public/application/components/data_grid/data_grid_table_columns.tsx
Comment (ashwin-pc) [inline src/plugins/discover/public/application/components/data_grid/data_grid_table_columns.tsx:1]: The same goes for this. Unit tests will be useful here

File: src/plugins/discover/public/application/components/data_grid/data_grid_table_context.tsx
Comment (ashwin-pc) [inline src/plugins/discover/public/application/components/data_grid/data_grid_table_context.tsx:19]: nit: 
```suggestion
export const DataGridContext = React.createContext<DataGridContextProps>(
  {} as DataGridContextProps
);

export const DiscoverGridContextProvider = DataGridContext.Provider;
export const useDataGridContext = () => React.useContext(DataGridContext);
```

File: src/plugins/discover/public/application/components/data_grid/data_grid_table_flyout.tsx
Comment (ashwin-pc) [inline src/plugins/discover/public/application/components/data_grid/data_grid_table_flyout.tsx:82]: this should come from `doc_view_links` registry

File: src/plugins/discover/public/application/components/utils/use_pagination.ts
Comment (ashwin-pc) [inline src/plugins/discover/public/application/components/utils/use_pagination.ts:1]: since this is also complete lets add unit tests for this

File: src/plugins/discover/public/application/view_components/canvas/discover_table.tsx
Comment (ashwin-pc) [inline src/plugins/discover/public/application/view_components/canvas/discover_table.tsx:23]: nit: I usually have my case sensitivity turned on so i dont want to accidentally miss this
```suggestion
  // TODO: get id from data explorer since it is handling the routing logic
```

File: src/plugins/discover/public/application/view_components/canvas/discover_table_app.tsx
Comment (ashwin-pc) [inline src/plugins/discover/public/application/view_components/canvas/discover_table_app.tsx:33]: most of these are already done in my side panel PR so i'd hold off implementing this until then :)

File: src/plugins/discover/public/application/view_components/canvas/discover_table_app.tsx
Comment (ashwin-pc) [inline src/plugins/discover/public/application/view_components/canvas/discover_table_app.tsx:40]: Do we need `EuiPage` and `EuiPageBody`? We already have them at the app root level

File: src/plugins/discover/public/application/view_components/canvas/discover_table_service.tsx
Comment (ashwin-pc) [inline src/plugins/discover/public/application/view_components/canvas/discover_table_service.tsx:19]: Do we need this? Right now the React tree looks like:

`Canvas` -> `DiscoverTable` -> `DiscoverTableService` -> `DiscoverTableApplication` -> `DataGrid`

Cant we instead simplify this to just be 

`Canvas` -> `DiscoverTable` -> `DataGrid`?

File: src/plugins/discover/public/application/view_components/canvas/utils/use_discover_canvas_service.ts
Comment (ashwin-pc) [inline src/plugins/discover/public/application/view_components/canvas/utils/use_discover_canvas_service.ts:1]: All files inside utils need to have unit tests. I wont hold back the PR if they arent there, but before we merge to main, we definitely need to add them

**Review by ashwin-pc — COMMENTED** [review]
The PR looks good to me, just one blocking change to add the `doc_viewer_links` service back. Unit tests would be really appreciated since we will need them before merge to main anyways.

File: src/plugins/discover/public/application/components/data_grid/constants.ts
Comment (ananzh) [reply src/plugins/discover/public/application/components/data_grid/constants.ts:6]: will change

File: src/plugins/discover/public/application/doc_views_links/doc_views_links_registry.ts
Comment (ananzh) [reply src/plugins/discover/public/application/doc_views_links/doc_views_links_registry.ts:1]: will add it back

File: src/plugins/discover/public/application/components/data_grid/data_grid_table_cell_value.tsx
Comment (ananzh) [reply src/plugins/discover/public/application/components/data_grid/data_grid_table_cell_value.tsx:1]: Will add an initial unit test for now. This [data_grid_table_cell_value.tsx](https://github.com/opensearch-project/OpenSearch-Dashboards/pull/4564/files/f3d17ed1dde77000a8b13e2ba783fedd5a655a45#diff-072b438f64a2ae2a0820a1763aa55440e95ebc612360230d2db1d8f1a0a6de1f) will need further polish to include nested fields and others.

File: src/plugins/discover/public/application/components/data_grid/data_grid_table_columns.tsx
Comment (ananzh) [reply src/plugins/discover/public/application/components/data_grid/data_grid_table_columns.tsx:1]: will add a unit test

File: src/plugins/discover/public/application/components/data_grid/data_grid_table_flyout.tsx
Comment (ananzh) [reply src/plugins/discover/public/application/components/data_grid/data_grid_table_flyout.tsx:82]: will modify this part later.

File: src/plugins/discover/public/application/components/utils/use_pagination.ts
Comment (ananzh) [reply src/plugins/discover/public/application/components/utils/use_pagination.ts:1]: currently many params used in pagnation, like pageSize, are fixed. later, we will make it configurable. but, sure, let me add an initial test.

File: src/plugins/discover/public/application/view_components/canvas/discover_table.tsx
Comment (ananzh) [reply src/plugins/discover/public/application/view_components/canvas/discover_table.tsx:23]: will replace all ToDo to TODO

File: src/plugins/discover/public/application/view_components/canvas/discover_table_app.tsx
Comment (ananzh) [reply src/plugins/discover/public/application/view_components/canvas/discover_table_app.tsx:40]: will remove

File: src/plugins/discover/public/application/view_components/canvas/discover_table_service.tsx
Comment (ananzh) [reply src/plugins/discover/public/application/view_components/canvas/discover_table_service.tsx:19]: This DiscoverTableApplication actually need to rename. It will include DiscoverChart (HitsCount, TimeChartHeader and Histogram) and DiscoverTable.
This discover_table_service is used to have any updates on services, like data & saved search for both DiscoverChart and DiscoverTable. 

We could re-evaluate and simply after canvas parts all implemented.

File: src/plugins/discover/public/application/view_components/canvas/utils/use_discover_canvas_service.ts
Comment (ananzh) [reply src/plugins/discover/public/application/view_components/canvas/utils/use_discover_canvas_service.ts:1]: Yes and will do. Currently, use_discover_canvas_service is not done. We will add tests for every utility functions after canvas is done.

**Comment by ananzh** [comment]
@ashwin-pc fixed all your comments. Some major changes are 1) add some initial tests. I put index pattern mock into a common mock folder bc it will be used by many other comps. Currently `dataRowsMock` is in each test file, but if there is more duplication, we will also move it to common folder. 2) add doc_views_link back. will see how to use it better and verify these links work in later work.

File: src/plugins/discover/public/__mock__/index_pattern_mock.ts
Comment (ashwin-pc) [inline src/plugins/discover/public/__mock__/index_pattern_mock.ts:1]: Do we need this? Cant we use `import { getStubIndexPattern } from '../../../../../../data/public/test_utils';` like the one found in `src/plugins/discover/public/application/components/sidebar/lib/field_calculator.test.ts`?

**Review by ashwin-pc — APPROVED** [review]
Wont block on the new index pattern mock, but in a followup PR can you replace the custom mock with an existing one so that we dont have many ways to mock the same thing. Index patterns may change soon and custom mocks like this will make it annoying to update tests that rely on them.

File: src/plugins/discover/public/__mock__/index_pattern_mock.ts
Comment (ananzh) [reply src/plugins/discover/public/__mock__/index_pattern_mock.ts:1]: Yes we could. This is a great finding. I think we just need to add a new data in fixture. Let me do it in a follow up PR.

**Pull request merged**
